test/function.localstack.test.ts

```diff
@@ -0,0 +1,533 @@
+    name,
+    func,
+    expected,
+    payload
+  ) => {
+    testResource(
+      name,
+      (parent) => {
+        const res = func(parent);
+        const [funcRes, outputs] =
+          res instanceof Function ? [res, {}] : [res.func, res.outputs];
+        return {
+          outputs: {
+            function: funcRes.resource.functionName,
+            ...outputs,
+          },
+        };
+      },
+      async (context) => {
+        const exp =
+          // @ts-ignore
+          typeof expected === "function" ? expected(context) : expected;
+        // @ts-ignore
+        const pay = typeof payload === "function" ? payload(context) : payload;
+        await testFunction(context.function, pay, exp);
+      }
+    );
+  };
+
+  testFunctionResource.skip = (name, _func, _expected, _payload?) =>
+    testResource.skip(
+      name,
+      () => {},
+      async () => {}
+    );
+
+  testFunctionResource(
+    "Call Lambda",
+    (parent) => {
+      return new Function(parent, "func2", async (event) => event);
+    },
+    {}
+  );
+
+  testFunctionResource(
+    "Call Lambda from closure",
+    (parent) => {
+      const create = () =>
+        new Function(parent, "function", async (event) => event);
+
+      return create();
+    },
+    {}
+  );
+
+  testFunctionResource(
+    "Call Lambda from closure with variables",
+    (parent) => {
+      const create = () => {
+        const val = "a";
+        return new Function(parent, "function", async () => val);
+      };
+
+      return create();
+    },
+    "a"
+  );
+
+  testFunctionResource(
+    "Call Lambda from closure with parameter",
+    (parent) => {
+      const create = (val: string) => {
+        return new Function(parent, "func5", async () => val);
+      };
+
+      return create("b");
+    },
+    "b"
+  );
+
+  const create = (parent: Construct, id: string, val: string) => {
+    return new Function(parent, id, async () => val);
+  };
+
+  testFunctionResource(
+    "Call Lambda from closure with parameter multiple 1",
+    (parent) => create(parent, "func6", "c"),
+    "c"
+  );
+
+  testFunctionResource(
+    "Call Lambda from closure with parameter multiple 2",
+    (parent) => create(parent, "func7", "d"),
+    "d"
+  );
+
+  testFunctionResource(
+    "Call Lambda with object",
+    (parent) => {
+      const create = () => {
+        const obj = { val: 1 };
+        return new Function(parent, "function", async () => obj.val);
+      };
+
+      return create();
+    },
+    1
+  );
+
+  testFunctionResource(
+    "Call Lambda with math",
+    (parent) =>
+      new Function(parent, "function", async () => {
+        const v1 = 1 + 2; // 3
+        const v2 = v1 * 3; // 9
+        return v2 - 4; // 5
+      }),
+    5
+  );
+
+  testFunctionResource(
+    "Call Lambda payload",
+    (parent) =>
+      new Function(parent, "function", async (event: { val: string }) => {
+        return `value: ${event.val}`;
+      }),
+    "value: hi",
+    { val: "hi" }
```

Comment:
so just so i understand, the contract is:
1. create the Function construct
2. expected output
3. input argument

Reply:
For users, we're going to need this to be:
```ts
test("blah", async () => {
  const func = new Function(..);
  expect(await func(input)).toEqual(expected);
});
```

Reply:
#131 covers running the closure of integrations as typescript.

The tests here look funky because of the integration with localstack and actually running the lambda on localstack. This is how we'd test any CDK resource, not specific to our function.

Did you have an idea how testing would be done by developer? I don't think we'd want to integrate localstack in official, so its up to the developer how they'd test resources running again AWS or localstack (which look like any other CDK resource).

I guess #131 is the closest to what developers can do today, just running their lambda code, but without the step of serialization and integration being the unknown variables. 

Another option is to expose the serialized form somehow to run locally to catch any issues... not sure yet.